### PR TITLE
[Operator] Add batch_norm

### LINKED ISSUE #258: Code Contribution: 【Lv3】【Operator Development】batch_norm
将以下算子贡献至https://github.com/FlagOpen/FlagGems/tree/master/src/flag_gems/ops 目录下

batch_norm

认领方式：
1. Issue中留言评论：「认领人名称+认领」
2. 在贡献者微信群内发送：认领的Issue号例如：「认领+[https://github.com/FlagOpen/FlagGems/issues/xxx]」

### LINKED ISSUE #315: Code Contribution: 【Lv3】【Operator Development】batch_norm_backward
## Description 任务介绍
Develop backward function for batch_norm operator.
开发batch_norm算子的反向功能。
## Requirements 任务要求
__Interface 接口__
`batch_norm_backward(Tensor grad_out, Tensor input, Tensor weight, Tensor? running_mean, Tensor? running_var, Tensor? save_mean, Tensor? save_var, bool update, float eps, bool[3] output_mask, Tensor reserve) -> (Tensor, Tensor, Tensor)`
__Function reference 功能参考__
https://pytorch.org/docs/stable/generated/torch.nn.functional.batch_norm.html
__Implementation reference 实现参考__
https://github.com/FlagOpen/FlagGems/blob/master/src/flag_gems/ops/groupnorm.py
https://github.com/FlagOpen/FlagGems/blob/master/src/flag_gems/ops/layernorm.py

The operator should support all optional arguments defined in the interface and return the corresponding number of values. 
算子应支持接口中定义的所有参数选项，并返回相应数量的值。
## DDL 提交时间
Please submit a Pull Request within 3 weeks after accepting the assignment.
请于接取任务后三周内提交PR。
Please provide both accuracy test and performance test code.
请同时提供实现正确性测试与性能测试代码。

### PR DESCRIPTION
**Pull request by 2niuhe**
### PR Category
Operator

### Type of Change
New Feature

### Description
Implement batch_norm operator

- forward
- backward

### Issue

- Resolves #258 
- Resolves #315

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [ ] Change is fully covered by a UT.

### CODE REVIEW

File: src/flag_gems/ops/batch_norm.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+@libentry()
+@triton.autotune(
+    configs=runtime.get_triton_config("batch_norm"),
+    key=["batch_dim", "spatial_dim"],
+    restore_value=["running_mean_pointer", "running_var_pointer"],
+)
+@triton.heuristics(
+    {
+        "BLOCK_SIZE_BATCH": lambda args: next_power_of_2(args["batch_dim"]),
+        "BLOCK_SIZE_SPATIAL": BLOCK_SIZE_SPATIAL_heuristic,
```
Comment (StrongSpoon) [inline src/flag_gems/ops/batch_norm.py]: we are working on changing one-tile algorithm into loop tiling. please refer to max/min/lof_softmax and update the strategy. ;)

File: benchmark/test_norm_perf.py [hunk 2]
```diff
@@ -73,6 +73,20 @@ def instancenorm_input_fn(shape, dtype, device):
...
+    inp = torch.randn(shape, dtype=dtype, device=device)
+    weight = torch.randn((C,), dtype=dtype, device=device)
+    bias = torch.randn((C,), dtype=dtype, device=device)
+    running_mean = torch.randn((C,), dtype=dtype, device=device)
+    running_var = torch.randn((C,), dtype=dtype, device=device)
+    training = True
+    momentum = 0.1
+    eps = 1e-5
+    cudnn_enabled = True
+    yield inp, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled
```
Comment (StrongSpoon) [inline benchmark/test_norm_perf.py:87]: I suggest yielding inp, weight and bias in core mode and yielding them all in comprehensive mode.

File: src/flag_gems/ops/batch_norm.py (hunk 1 above)
Comment (2niuhe) [reply src/flag_gems/ops/batch_norm.py]: Assuming the input  to batch_norm has the shape **(batch, channel, spatial)**.

The implementation of batch_norm **is already based on loop tiling, where tiling occurs along the spatial dimension** while fully loading the batch dimension. This approach differs from the operators mentioned, such as max, min, which only require computations along a specified dimension without involving others. In the case of batch_norm, both the batch and spatial dimensions need to be fully loaded.

My previous consideration was that introducing tiling along the batch dimension would result in a **nested loop structure**, which may not be meaningful when batch is not large.

You might be suggesting that applying loop tiling on the batch dimension could indeed improve performance when the batch size is large, as it would allow for more contiguous memory access in the spatial dimension. For instance, if the batch size is 16,384 and the spatial size loaded per loop iteration is only 1, this could lead to inefficient memory access patterns. **I will try implementing loop tiling on both the batch and spatial dimensions**.

File: src/flag_gems/ops/batch_norm.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+            count += curr_count
+
+            prev_mean = mean
+            mean += (tl.sum(curr_input) - curr_count * mean) / count
+            deltas = tl.where(
+                batch_mask[:, None] & spatial_mask[None, :],
+                (curr_input - mean) * (curr_input - prev_mean),
+                0.0,
+            )
+            var += tl.sum(deltas)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/batch_norm.py]: this method cannot fully utilize vectorization/tensorization and leads to sequential computation.

File: src/flag_gems/ops/batch_norm.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+            )
+            curr_input = tl.load(
+                curr_input_pointer, mask=batch_mask[:, None] & spatial_mask[None, :]
+            ).to(tl.float32)
+
+            spatial_count = min(
+                BLOCK_SIZE_SPATIAL, spatial_dim - block_ind * BLOCK_SIZE_SPATIAL
+            )
+            curr_count = spatial_count * batch_dim
+            count += curr_count
```
Comment (StrongSpoon) [inline src/flag_gems/ops/batch_norm.py]: we can set reasonable default value of tl.load to avoid keeping counter.

File: src/flag_gems/ops/batch_norm.py (hunk 1 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/batch_norm.py]: got it

File: src/flag_gems/ops/batch_norm.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+            * (curr_output_grad - (term1 * curr_pre_lin + term2) / count)
+        )
+        tl.store(
+            curr_input_grad_pointer,
+            curr_input_grad,
+            mask=batch_mask[:, None] & spatial_mask[None, :],
+        )
+
+        if affine:
+            weight_grad += tl.sum(curr_pre_lin * curr_output_grad)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/batch_norm.py]: ditto

File: src/flag_gems/ops/batch_norm.py [hunk 6]
```diff
@@ -0,0 +1,489 @@
...
+
+        if ctx.affine:
+            weight_grad = torch.empty((feat_dim,), device=input.device)
+            bias_grad = torch.empty_like(weight_grad)
+
+        else:
+            weight_grad = bias_grad = None
+
+        # Launches 1D grid where each program operates over one feature.
+        grid = lambda _: (feat_dim,)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/batch_norm.py]: grid hasn't to be a lambda expression. initialize it as a tuple.

**Comment by StrongSpoon** [comment]
please fix bug in accuracy test:)

File: tests/test_norm_ops.py [hunk 7]
```diff
@@ -475,3 +477,102 @@ def test_accuracy_vectornorm(shape, ord, dim, keepdim, dtype):
...
+        (16, 3),
+        (32, 32, 32),
+        (8, 32, 224, 224),
+        (2050, 16, 32, 32),
+        (8, 16, 3, 224, 224),
+    ],
+)
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES)
+@pytest.mark.parametrize("affine", [True, False])
+@pytest.mark.parametrize("require_grad", [True, False])
```
Comment (StrongSpoon) [inline tests/test_norm_ops.py:495]: just set requires_grad as True

**Review by StrongSpoon — APPROVED** [review]
LG

File: tests/test_norm_ops.py (hunk 7 above)
Comment (2niuhe) [reply tests/test_norm_ops.py:495]: Thanks for the suggestion. Some code paths require requires_grad to be False to execute, which helps us cover more execution paths in the tests.

Comment (StrongSpoon) [reply tests/test_norm_ops.py:495]: got it
